backend/apid/middlewares/request_info.go

```diff
@@ -0,0 +1,82 @@
+package middlewares
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/sensu/sensu-go/types"
+)
+
+// RequestInfo is an HTTP middleware that populates a context for the request,
+// to be used by other middlewares. You probably want this middleware to be
+// executed early in the middleware stack.
+type RequestInfo struct{}
+
+// Then is the RequestInfo middleware's main logic, heavily inspired by
+// Kubernetes' WithRequestInfo.
+//
+// The general, expected format of a path one of the following:
+// - /apis/{group}/{version}/namespaces, to create or list namespaces
+// - /apis/{group}/{version}/namespaces/{namespace}, to retrieve, update or
+//   delete namespace {namespace}
+// - /apis/{group}/{version}/namespaces/{namespace}/{type}, to create or list
+//   objects of type {type} in namespace {namespace}
+// - /apis/{group}/{version}/namespaces/{namespace}/{type}/{name}, to retrieve,
+//   update or delete object {name} of type {type} in namespace {namespace}
+//
+// This middleware tries to fill in as many fields as it can in the
+// types.RequestInfo struct added to the context, but there is no guarantee of
+// any fields being filled in.
+func (i RequestInfo) Then(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		info := types.RequestInfo{}
+
+		router := mux.NewRouter().PathPrefix("/apis/{group}/{version}/namespaces/").Subrouter()
+		router.Path("").Methods("GET", "HEAD", "POST")
+		router.Path("/{namespace}").Methods("GET", "HEAD", "PUT", "DELETE")
+		router.Path("/{namespace}/{type}").Methods("GET", "HEAD", "POST")
+		router.Path("/{namespace}/{type}/{name}").Methods("GET", "HEAD", "PUT", "DELETE")
```

Comment:
We also need to support resources that are within namespaces (e.g. users).

Reply:
That.

Also, I now have a question about these router paths.

I don't think you should be able to GET/HEAD/POST on `/apis/group/version/namespaces`. Was the intent to allow people to easily list everything in all namespaces? I could maybe agree with that, but let's not allow POST here. Let's require API clients to fully qualify the path of the thing they want to create.

As far as GET for this route goes, I'd almost rather API clients had to enumerate the namespaces they had access to and then query all objects of a given type across those namespaces. It'd be a heavier query, but I'd rather put that burden on API clients and not the apiserver.

The same goes for `/apis/group/version/namespaces/namespace`. GET would list all objects for that group in the given namespace at the specified version. Again, I'd rather people had to query specific types instead of us having to iterate through all known types for that version and that group. I don't see this being a compelling use case to support this route.

Reply:
Also, you certainly shouldn't be able to DELETE on `/apis/group/version/namespaces/namespace`. I'm not even sure what that means.

Reply:
I think what we actually want to support is:

`GET, HEAD, POST /apis/group/version/type` :: operate on a set of a versioned non-namespaced type.

`GET, HEAD, POST /apis/group/version/namespaces/namespace/type` :: operate on a set of a versioned namespaced type.

`GET, HEAD, PUT, DELETE /apis/group/version/type/name` :: operate on a versioned non-namespaced object of a specified type

`GET, HEAD, PUT, DELETE /apis/group/version/namespaces/namespace/type/name` :: operate on a versioned namespaced object of a specified type.

Do we actually need any other routes than those?

Reply:
> Also, you certainly shouldn't be able to DELETE on `/apis/group/version/namespaces/namespace`. I'm not even sure what that means.

How will people delete namespaces?

Reply:
We have to have a namespace API much like we have an API for organizations and environments now.

Reply:
Which reminds me, we need to validate namespaces when they're provided by the user in input.

Reply:
> I don't think you should be able to GET/HEAD/POST on
> `/apis/group/version/namespaces`. Was the intent to allow people to easily
> list everything in all namespaces?

The intent is to be able to list the available namespaces (GET) and create a new
namespace (POST).

`Namespace`, [defined here](https://github.com/sensu/sensu-go/blob/devel/api/internal/apis/core/namespace.go),
appears to be just another object. So a GET on `/namespaces` is a list verb
internally and would return a list of `Namespace` objects, but not all the
objects that are under all the namespaces. A POST on `/namespaces` expects a
`Namespace` object in the request body, which would hold the name of the new
namespace that needs to be created.

> I could maybe agree with that, but let's not allow POST here. Let's require
> API clients to fully qualify the path of the thing they want to create.

If that's also what we do with all the other object types, then yes. If not, I
would disagree and err on the side of consistency for all object types.

My understanding was that, for example, to create a `Check` object, one would
POST the object, with its name in its metadata, to `{...}/checks`. That would
result in an object being available at `{...}/checks/{name}`. I thought all
objects would work the same.

> The same goes for `/apis/group/version/namespaces/namespace`. GET would list
> all objects for that group in the given namespace at the specified version.
> Again, I'd rather people had to query specific types instead of us having to
> iterate through all known types for that version and that group. I don't see
> this being a compelling use case to support this route.

Based on my answer above, GET on
`/apis/{group}/{version}/namespaces/{namespace}`
would only return a single `Namespace` object, not everything that's in the
namespace. It doesn't sound very useful, and it's mostly for consistency with
all the other object types for which a GET on a specific object name gives you
back that object.

> Also, you certainly shouldn't be able to DELETE on /apis/group/version/namespaces/namespace. I'm not even sure what that means.

An HTTP DELETE on `/apis/{group}/{version}/namespaces/{namespace}` deletes the
`Namespace` object with name `{namespace}`.

Reply:
You don't manage Namespaces within, for example, the RBAC API. Namespaces are global objects that are defined via a namespace API (assuming it is in core) `/apis/core/{version}/namespace`. Consequently, "list"-ing namespaces is done via the namespace API, not via `/apis/{group}/{version}/namespaces`.

My comment about POST above is that if you want to create an object of a specific type, you would `POST /apis/{group}/{version}/{type}`. 

These behaviors should be consistent across all object types.

Again, to be clear, interacting with `Namespace` objects only happens within the core API where the `Namespace` object is defined.

As a general rule of thumb: you can only modify objects that belong to an API group _within_ that API group.

Reply:
To be very clear, the only ways to interact with `Namespace` objects are:

`GET /apis/core/v1/namespace` - list namespaces
`GET /apis/core/v1/namespace/{name}` - get a specific namespace
`HEAD /apis/core/v1/namespace/{name}` - validate the existence of a namespace
`POST /apis/core/v1/namespace` - create a new namespace
`DELETE /apis/core/v1/namespace` - delete a namespace (except you cannot delete the default namespace

I don't even think we should allow PUT to namespaces.